src/WebsocketServer.js

```diff
@@ -95,100 +108,107 @@ module.exports = class WebsocketServer extends events.EventEmitter {
-                    request.streamPartition,
-                    request.resendOptions.resend_from_time,
-                    msgHandler,
-                    doneHandler,
-                )
-            } else {
-                debug('handleResendRequest: unknown resend request: %o', JSON.stringify(request))
-                connection.send(new Protocol.ErrorResponse(`Unknown resend options: ${JSON.stringify(request.resendOptions)}`))
-            }
+        this.streamFetcher.authenticate(request.streamId, request.apiKey, request.sessionToken).then(() => {
+            const streamingStorageData = resendTypeHandler()
+            streamingStorageData.on('data', msgHandler)
+            streamingStorageData.on('end', doneHandler)
         }).catch((err) => {
-            connection.send(new Protocol.ErrorResponse(`Failed to request resend from stream ${
+            connection.sendError(`Failed to request resend from stream ${
                 request.streamId
             } and partition ${
                 request.streamPartition
-            }: ${err.message}`))
+            }: ${err.message}`)
         })
     }
 
+    handleResendLastRequest(connection, request) {
+        this.handleResendRequest(connection, request, () => this.storage.fetchLatest(
+            request.streamId,
+            request.streamPartition,
+            request.numberLast,
+        ))
+    }
+
+    handleResendFromRequest(connection, request) {
+        if (request.publisherId) {
+            this.handleResendRequest(connection, request, () => this.storage.fetchFromMessageRefForPublisher(
+                request.streamId,
+                request.streamPartition,
+                request.fromMsgRef,
+                request.publisherId,
+            ))
+        } else {
+            this.handleResendRequest(connection, request, () => this.storage.fetchFromTimestamp(
+                request.streamId,
+                request.streamPartition,
+                request.fromMsgRef.timestamp,
+            ))
+        }
+    }
+
+    handleResendRangeRequest(connection, request) {
+        if (request.publisherId) {
+            this.handleResendRequest(connection, request, () => this.storage.fetchBetweenMessageRefsForPublisher(
+                request.streamId,
+                request.streamPartition,
+                request.fromMsgRef,
+                request.toMsgRef,
+                request.publisherId,
+            ))
+        } else {
+            this.handleResendRequest(connection, request, () => this.storage.fetchBetweenTimestamps(
+                request.streamId,
+                request.streamPartition,
+                request.fromMsgRef.timestamp,
+                request.toMsgRef.timestamp,
+            ))
+        }
+    }
+
+    /* eslint-disable class-methods-use-this */
+    handleResendRequestV0(connection) {
+        connection.sendError('ResendRequestV0 is not supported anymore. Please update your Streamr client.')
```

Comment:
- Resend last requests should still be ok without any problems?
- If we fake `kafkaOffset = timestamp` for `V0` clients, then would resend from/range requests translate to resend from/range timestamp requests? I'm not sure if it would cause mayhem on old clients? Not if timestamps happen to be unique I guess (in many cases they might be).